src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+// are enough types that we need to flush to multiple events).
+static
+void
+ep_rt_bulk_type_value_clear (BulkTypeValue *bulk_type_value)
+{
+	bulk_type_value->fixed_sized_data.type_id = 0;
+	bulk_type_value->fixed_sized_data.module_id = 0;
+	bulk_type_value->fixed_sized_data.type_name_id = 0;
+	bulk_type_value->fixed_sized_data.flags = 0;
+	bulk_type_value->fixed_sized_data.cor_element_type = 0;
+	bulk_type_value->c_type_parameters = 0;
+	memset (bulk_type_value->rg_type_parameters, 0, K_SIZE_OF_TYPE_PARAMETER_ARRAY);
+	memset (bulk_type_value->rg_mono_type_parameters, 0, K_SIZE_OF_TYPE_PARAMETER_ARRAY);
+	bulk_type_value->s_name = NULL;
+}
+
+static
+int
+ep_rt_mono_get_byte_count_in_event (BulkTypeValue *bulk_type_value)
+{
+	int s_name_len = 0;
+
+	return sizeof (bulk_type_value->fixed_sized_data.type_id) + 	// Fixed Sized Data
+		sizeof (bulk_type_value->fixed_sized_data.module_id) +
+		sizeof (bulk_type_value->fixed_sized_data.type_name_id) +
+		sizeof (bulk_type_value->fixed_sized_data.flags) +
+		sizeof (bulk_type_value->fixed_sized_data.cor_element_type) +
+		sizeof (bulk_type_value->c_type_parameters) +		// Type parameters
+		(s_name_len + 1) * sizeof (ep_char8_t) +		// Size of name, including null terminator
+		bulk_type_value->c_type_parameters * sizeof (uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+#define CB_MAX_ETW_EVENT (63 * 1024)
+
+// The maximum event size, and the size of the buffer that we allocate to hold the event contents.
+#define K_SIZE_OF_EVENT_BUFFER 65536
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array. (Intentionally overestimate the size of the non-array parts to keep it safe.)
+// This follows CoreCLR's kMaxBytesTypeValues.
+#define K_MAX_BYTES_TYPE_VALUES (CB_MAX_ETW_EVENT - 0x30)
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+#define K_MAX_COUNT_TYPE_VALUES ((uint32_t)(128 - 2) / 4)
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values [K_MAX_COUNT_TYPE_VALUES];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer;
```

Comment:
If you are not depending on order of fields, then you should move this field just after m_rg_bulk_type_values to reduce padding.